src/helpers/gateway/send.rs

```diff
@@ -69,13 +71,24 @@ impl GatewaySender {
         // TODO: make OrderingSender::send fallible
         // TODO: test channel close
         let i = usize::from(record_id);
+        self.pending_records.lock().unwrap().insert(i);
```

Comment:
I think it is actually harder than that. Protocols may hang even before they call send and this approach does not seem to let us detect that. 

If we look at send buffer, it is essentially a contiguous region in memory that is capable of storing `N` homogenious messages [_, _, _, _, _, _]. There is a pending window of records we want to send at each step.

For example, if buffer size is 10 messages x 4 bytes each, OrderingSender will have a buffer of size 40 and first will accept records 0..9. Once it receives them all, it flushes it down to transport and moves the window to recrods 10..19, etc.

Our job here is to figure out the current window (start, end) and figure out the gaps inside that buffer. Gaps are records that are never attempted to be sent (not inside `State`/`waiting`) and then print them

Ideal output would be something like this:

```bash
step: A, waiting to send: 14, 17
step B, waiting to send: 3, 7, 8
step C: waiting to receive: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9